### ISSUE TITLE
Invalid read after `setlocale`

### ISSUE THREAD

**Issue by YannickJadoul** [eliaskosunen/scnlib#35]
I've just tracked down an invalid read with ASAN ("heap-use-after-free"), to the `setlocale` calls in `scn::detail::read_float_impl`. E.g.:

https://github.com/eliaskosunen/scnlib/blob/0a2a4ba2551deaebedd1971519f8f0c33dfed539/src/reader.cpp#L92-L97

It's not entirely clear from [cppreference.com](https://en.cppreference.com/w/cpp/locale/setlocale), but I believe that the value of `setlocale("LC_NUMERIC", nullptr)` is invalidated after `std::setlocale(LC_NUMERIC, "C")`. It somehow only made my Windows CI tests crash, but is pretty consistent there, and ASAN also triggers it on Linux.

ASAN stops complaining when I make the following change:

```diff
@@ -62,12 +62,12 @@ namespace scn {
             static expected<float> get(const wchar_t* str, size_t& chars)
             {
                 wchar_t* end{};
-                const auto loc = std::setlocale(LC_NUMERIC, nullptr);
+                const auto loc = std::string(std::setlocale(LC_NUMERIC, nullptr));
                 std::setlocale(LC_NUMERIC, "C");
                 errno = 0;
                 float f = std::wcstof(str, &end);
                 chars = static_cast<size_t>(end - str);
-                std::setlocale(LC_NUMERIC, loc);
+                std::setlocale(LC_NUMERIC, loc.c_str());
                 if (errno == ERANGE) {
                     errno = 0;
                     return error(error::value_out_of_range,
```

If you wish, I could make a PR to fix this?

---

That being reported, a related question:
One of the reasons I am very keen on using `scn` is because it provides locale-independent float-to-string conversions, and I had hoped it would stay away from the ugly global `setlocale` state. So kind of related to #24, would you be interested in getting rid of these `setlocale` calls?

I have a [bit of a hackish solution for the moment, tapping into the low-level parts of iostream](https://github.com/YannickJadoul/Parselmouth/blob/6bea71de4ee345782ddc9ea06b1586765d0ef287/praat/melder/melder_atof.cpp#L26-L52), but it does manage to not call `setlocale`. And [afborchert/fmt-scan](https://github.com/afborchert/fmt-scan) seems to be doing [something similar](https://github.com/afborchert/fmt-scan/blob/master/scan.hpp#L1049-L1056) (though probably much better than me). Maybe such a solution can be hidden inside scnlib? Compared to the suggestions in #24, these wouldn't require external dependencies.

**Comment by eliaskosunen** [comment]
The first part of the report resolved in 756a65d per your suggestion. I'll need to look into the second part (not using C locale when parsing floats) at some point.

I elected for not requesting a PR from you, as this issue has already been open for long enough. My military service is really limiting my opportunities to maintain this project, so sorry about the delay.

**Comment by YannickJadoul** [comment]
No problem at all! I fully understand that you working for free and maintaining an open source project is not your first priority.
Still, thanks for doing it, though! :-)

If you do want a hand on the second part, do ping me if you want.
